kong/db/schema/others/declarative_config.lua

```diff
@@ -506,6 +507,75 @@ local function validate_references(self, input)
 end
 
 
+function DeclarativeConfig.validate_references_full(self, input)
+  return validate_references(self, input)
+end
+
+
+function DeclarativeConfig.validate_references_sync(deltas, deltas_map)
+  local errs = {}
+
+  for _, delta in ipairs(deltas) do
+    local item_type = delta.type
+    local item = delta.entity
+    local ws_id = delta.ws_id or kong.default_workspace
+
+    local foreign_refs = foreign_references[item_type]
+
+    if not item or item == null or not foreign_refs then
+      goto continue
+    end
+
+    for k, v in pairs(item) do
+
+      -- Try to check if item's some foreign key exists in the deltas or LMDB.
+      -- For example, `item[k]` could be `<router_entity>["service"]`, we need
+      -- to find the referenced foreign service entity for this router entity.
+
+      local foreign_entity = foreign_refs[k]
+
+      if foreign_entity and v ~= null then  -- k is foreign key
+
+        local db = kong.db[foreign_entity]
+
+        -- try to find it in deltas
+        local pks = DeclarativeConfig.pk_string(db.schema, v)
+        local fvalue = deltas_map[pks]
+        if fvalue then
+          goto found
+        end
+
+        -- try to find it in DB (LMDB)
+        fvalue, _ = db:select(v, { workspace = ws_id })
+
+        -- could not find its foreign reference
+        if not fvalue then
+          errs[item_type] = errs[item_type] or {}
+          errs[item_type][foreign_entity] = errs[item_type][foreign_entity] or {}
+
+          local msg = fmt("could not find %s's foreign refrences %s (%s)",
+                          item_type, foreign_entity,
+                          (type(v) == "string" and v or cjson_encode(v)))
+
+          insert(errs[item_type][foreign_entity], msg)
+        end
+      end
+
+      ::found::
+    end
+
+    ::continue::
+  end
+
+
+  if next(errs) then
```

Comment:
Could it be `table.isempty`?

Reply:
i think next is ok, this code comes from other similar part of handing errs in kong.